### PR TITLE
Adding storage fields to accounts

### LINKED ISSUE #421: Add storage capacity and usage fields to accounts
## Issue To Be Solved

To enforce storage limits in accounts it is necessary to query storage usage information.

## Suggest A Solution

- [ ] Add a two new read-only fields to accounts (`AuthAccount` and `PublicAccount`):
  - `let storageCapacity: UInt64`: the maximum amount of storage that the account may store
  - `let storageUsed: UInt64`: the currently used amount of storage
    - TBD: Does this return the current amount or when the transaction started?
- [ ] Extend the runtime interface:
  ```go
  // GetStorageUsed gets storage used in bytes by the address at the moment of the function call.
  GetStorageUsed(address Address) (value uint64, err error)
  // GetStorageCapacity gets storage capacity in bytes on the address.
  GetStorageCapacity(address Address) (value uint64, err error)
  ```
- [x] Tests
- [x] Documentation

### PR DESCRIPTION
**Pull request by janezpodhostnik**
Closes #421 

## Description

see #421 
______

For contributor use:

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/value.go [hunk 1]
```diff
@@ -6380,22 +6380,27 @@ type AccountValue interface {
 }
 
 // AuthAccountValue
-
 type AuthAccountValue struct {
 	Address                 AddressValue
+	StorageUsedGet          func() UInt64Value
+	StorageCapacityGet      func() UInt64Value
```
Comment (turbolent) [inline runtime/interpreter/value.go:6386]: maybe make those unexported, as they are an implementation detail

File: runtime/tests/interpreter/interpreter_test.go [hunk 2]
```diff
@@ -6615,6 +6615,8 @@ func TestInterpretContractAccountFieldUse(t *testing.T) {
 						return map[string]interpreter.Value{
 							"account": interpreter.NewAuthAccountValue(
 								addressValue,
+								func() interpreter.UInt64Value { return interpreter.UInt64Value(0) },
```
Comment (turbolent) [inline runtime/tests/interpreter/interpreter_test.go:6618]: Maybe refactor this to a function declaration, e.g. and replace the line here and in the other cases below:

```go
func returnZero() interpreter.UInt64Value {
	return interpreter.UInt64Value(0)
}
```

```suggestion
								returnZero,
```

File: runtime/runtime.go [hunk 3]
```diff
@@ -267,6 +267,8 @@ func (r *interpreterRuntime) newAuthAccountValue(
 ) interpreter.AuthAccountValue {
 	return interpreter.NewAuthAccountValue(
 		addressValue,
+		func() interpreter.UInt64Value { return interpreter.UInt64Value(runtimeInterface.GetStorageUsed(addressValue.ToAddress())) },
+		func() interpreter.UInt64Value { return interpreter.UInt64Value(runtimeInterface.GetStorageCapacity(addressValue.ToAddress())) },
```
Comment (turbolent) [inline runtime/runtime.go:271]: Maybe add line breaks to make this a bit more readable, same below

**Comment by turbolent** [comment]
Great work, this looks good! 👏 👍

File: runtime/sema/errors.go [hunk 4]
```diff
@@ -615,7 +615,7 @@ type InvalidVariableKindError struct {
 
 func (e *InvalidVariableKindError) Error() string {
 	if e.Kind == ast.VariableKindNotSpecified {
-		return fmt.Sprintf("missing variable kind")
+		return "missing variable kind"
```
Comment (janezpodhostnik) [inline runtime/sema/errors.go:618]: this is a linter error fix. It creeped in from somewhere else.

File: runtime/runtime.go [hunk 5]
```diff
@@ -958,6 +958,19 @@ func (r *interpreterRuntime) newCreateAccountFunction(
 		return trampoline.Done{Result: account}
 	}
 }
+func storageUsedGetFunction(addressValue interpreter.AddressValue, runtimeInterface Interface) func() interpreter.UInt64Value {
+	address := addressValue.ToAddress()
+	return func() interpreter.UInt64Value {
+		return interpreter.UInt64Value(runtimeInterface.GetStorageUsed(address))
```
Comment (turbolent) [inline runtime/runtime.go:964]: This call to the host environment should be wrapped with `wrapPanic`

File: runtime/runtime.go [hunk 6]
```diff
@@ -958,6 +958,19 @@ func (r *interpreterRuntime) newCreateAccountFunction(
...
+	}
+}
+
+func storageCapacityGetFunction(addressValue interpreter.AddressValue, runtimeInterface Interface) func() interpreter.UInt64Value {
+	address := addressValue.ToAddress()
+	return func() interpreter.UInt64Value {
+		return interpreter.UInt64Value(runtimeInterface.GetStorageCapacity(address))
```
Comment (turbolent) [inline runtime/runtime.go:971]: This call to the host environment should be wrapped with `wrapPanic`

File: docs/language/accounts.md [hunk 7]
```diff
@@ -370,3 +374,22 @@ let otherRef = authAccount.borrow<&{Other}>(from: /storage/counter)
...
+
+An account's storage used and storage capacity can be checked using the `storageUsed` and `storageCapacity` fields.
+The fields represent current values of storage which means this would be true:
+
+```
+let storageUsedBefore = authAccount.storageUsed
+authAccount.save(<-create Counter(count: 123), to: /storage/counter)
+let storageUsedAfter = authAccount.storageUsed
+
+return storageUsedBefore != storageUsedAfter // this is true
```
Comment (turbolent) [inline docs/language/accounts.md:393]:
```suggestion
let storageUsedChanged = storageUsedBefore != storageUsedAfter  // is true
```

File: docs/language/accounts.md [hunk 8]
```diff
@@ -370,3 +374,22 @@ let otherRef = authAccount.borrow<&{Other}>(from: /storage/counter)
...
+
+Accounts storage is limited by its storage capacity.
+
+At the end of every transaction all accounts storage used is compared to their storage capacity.
+If for any account its storage used is greater than its storage capacity the transaction fails.
+
+An account's storage used and storage capacity can be checked using the `storageUsed` and `storageCapacity` fields.
+The fields represent current values of storage which means this would be true:
+
+```
```
Comment (turbolent) [inline docs/language/accounts.md:388]:
```suggestion
```cadence
```

File: docs/language/accounts.md [hunk 9]
```diff
@@ -11,6 +11,8 @@ Every account can be accessed through two types:
   struct PublicAccount {
 
       let address: Address
+      let storageUsed: UInt8
+      let storageCapacity: UInt8
```
Comment (turbolent) [inline docs/language/accounts.md:15]:
```suggestion
      let storageUsed: UInt64
      let storageCapacity: UInt64
```

File: docs/language/accounts.md [hunk 10]
```diff
@@ -40,6 +42,8 @@ Every account can be accessed through two types:
   struct AuthAccount {
 
       let address: Address
+      let storageUsed: UInt8
+      let storageCapacity: UInt8
```
Comment (turbolent) [inline docs/language/accounts.md:46]:
```suggestion
      let storageUsed: UInt64
      let storageCapacity: UInt64
```

File: runtime/sema/type.go [hunk 11]
```diff
@@ -5382,6 +5382,14 @@ const accountTypeContractsFieldDocString = `
 The contracts of the account
 `
 
+const accountTypeStorageUsedFieldDocString = `
+The current amount of storage used on the account in bytes
```
Comment (turbolent) [inline runtime/sema/type.go:5386]:
```suggestion
The current amount of storage used by the account in bytes
```

File: runtime/sema/type.go [hunk 12]
```diff
@@ -5382,6 +5382,14 @@ const accountTypeContractsFieldDocString = `
...
+`
+
+const accountTypeStorageCapacityFieldDocString = `
+The storage capacity on the account in bytes
```
Comment (turbolent) [inline runtime/sema/type.go:5390]:
```suggestion
The storage capacity of the account in bytes
```

File: runtime/tests/checker/account_test.go [hunk 13]
```diff
@@ -1259,3 +1259,48 @@ func TestCheckAccount_getCapability(t *testing.T) {
...
+
+			t.Run(testName, func(t *testing.T) {
+
+				code := fmt.Sprintf(
+					`
+	                      fun test(): UInt64 {
+	                          return %s.%s
+	                      }
+
+                          let cap = test()
```
Comment (turbolent) [inline runtime/tests/checker/account_test.go:1290]:
```suggestion
                          let amount = test()
```

File: runtime/tests/checker/account_test.go [hunk 14]
```diff
@@ -1259,3 +1259,48 @@ func TestCheckAccount_getCapability(t *testing.T) {
...
+					fieldName,
+				)
+				checker, err := ParseAndCheckAccount(
+					t,
+					code,
+				)
+
+				require.NoError(t, err)
+				capType := checker.GlobalValues["cap"].Type
+				assert.Equal(t, &sema.UInt64Type{}, capType)
```
Comment (turbolent) [inline runtime/tests/checker/account_test.go:1302]:
```suggestion
				amountType := checker.GlobalValues["amount"].Type
				assert.Equal(t, &sema.UInt64Type{}, amountType)
```

File: runtime/sema/errors.go (hunk 4 above)
Comment (turbolent) [reply runtime/sema/errors.go:618]: Thanks for fixing this

**Review by turbolent — COMMENTED** [review]
Looks good, just a few minor suggestions

File: runtime/runtime.go (hunk 5 above)
Comment (janezpodhostnik) [reply runtime/runtime.go]: like this?

Comment (turbolent) [reply runtime/runtime.go]: yep, [`10f72f7` (#439)](https://github.com/onflow/cadence/pull/439/commits/10f72f7cce3d636982256b1682928f1728bd8d25#diff-bb8ad382f59edfae9297e446895a161036f39115189e67da9d9f377a684a8f82R965-R967) looks good!

**Review by turbolent — APPROVED** [review]
Nice!

**Comment by turbolent** [comment]
@joshuahannan could you please have a quick look at the docs? Thank you!

**Comment by janezpodhostnik** [comment]
@turbolent I'm sorry, but looking back at the issue and what i need in flow-go I remembered that I need the getStorage interface functions to also return an error.

Could you please check the latest commit.

**Comment by turbolent** [comment]
The interface is currently not very consistent: some functions do have an error return value, some don't, even though all of them could error in the host environment. So currently the functions that don't have an error return value panic themselves, and for functions that have an error return value, Cadence panics (like in your last commit). Either way is fine, the behaviour is the same. It would be nice to clean this up one day

**Comment by joshuahannan** [comment]
The documentation seems a little bare. Are we planning on adding a bunch of the details to the documentation that were in the FLIP? @janezpodhostnik

**Comment by turbolent** [comment]
@joshuahannan What else do you think should be added? 

I think the current amount is fine, and we should rather link to the Flow documentation about details for how storage fees work on Flow, I don't think this should be in the language documentation.

**Comment by joshuahannan** [comment]
Good point. It makes sense that you'd want to keep this documentation only focused exactly on what is directly relevant to Cadence and put the other docs somewhere else. I still think this needs more though. I'll make a PR that adds some things that I think would be useful for this

**Comment by turbolent** [comment]
@joshuahannan 

> I'll make a PR that adds some things that I think would be useful for this

Great, thank you!

**Pull request merged**